library/kotlin/src/io/envoyproxy/envoymobile/Client.kt

```diff
@@ -0,0 +1,73 @@
+package io.envoyproxy.envoymobile
+
+import java.nio.ByteBuffer
+
+interface Client {
+  fun startStream(request: Request, streamCallback: StreamCallback): StreamEmitter
+}
+
+interface StreamCallback {
+  fun onHeaders(statusCode: Int, headers: Map<String, List<String>>)
+  fun onData(byteBuffer: ByteBuffer, endStream: Boolean)
+  fun onMetadata(metadata: Map<String, List<String>>, endStream: Boolean)
+  fun onTrailers(trailers: Map<String, List<String>>)
+  fun onError(error: Error)
+
+  fun onCanceled() // TODO: onInterrupt()?
+  fun onCompletion()
+
+}
+
+interface StreamEmitter {
+  fun sendHeaders(headers: Map<String, List<String>>): Boolean
+  fun sendData(byteBuffer: ByteBuffer, endStream: Boolean): Boolean
+  fun sendMetadata(metadata: Map<String, List<String>>, endStream: Boolean): Boolean
+  fun sendTrailers(trailers: Map<String, List<String>>): Boolean
+
+  fun close(): Boolean
+  fun interrupt(): Boolean
+}
+
+class Error(
+    val code: Int,
+    val message: String,
+    val cause: Throwable? = null
```

Comment:
What is the intended use case for this field?

Reply:
Primarily wondering how it'd be used if we already have `code` and `message`

Reply:
This was primarily for exception surfacing -- I copied what Cronet does as they return this [exception](https://chromium.googlesource.com/chromium/src/+/lkgr/components/cronet/android/api/src/org/chromium/net/CronetException.java?autodive=0%2F):

```
/**
 * Base exception passed to {@link UrlRequest.Callback#onFailed UrlRequest.Callback.onFailed()}.
 */
public abstract class CronetException extends IOException {
    /**
     * Constructs an exception that is caused by {@code cause}.
     *
     * @param message explanation of failure.
     * @param cause the cause (which is saved for later retrieval by the {@link
     *         java.io.IOException#getCause getCause()} method). A null value is permitted, and
     *         indicates that the cause is nonexistent or unknown.
     */
    protected CronetException(String message, Throwable cause) {
        super(message, cause);
    }
}
```